### DISCUSSION TITLE
Allow MDX and other content formats to use `.md`

### DISCUSSION THREAD

**Discussion by bholmesdev** [withastro/roadmap#503] [Proposal]
### Body

# Summary

This proposal would allow `.md` support for formats like MDX and Markdoc.

# Background & Motivation

We have received several support forum issues asking for `.md` support from MDX since Astro 1.0 ([1](https://discord.com/channels/830184174198718474/1077511972163162182)). There are a few motivators:
- Ease of migration from other tools that allow `.md`, including Gatsby
- Compatibility with git-based CMSes that output to `.md` only

# Goals

- Allow integrations to "claim" the `.md` extension for use in content collections and ESM imports
- Allow integrations to override type defs for `.md` imports

**Comment by Princesseuh** [comment]
I totally get the usage, but I'm personally minus one on this. I think it is fundamentally wrong to fake a file extension. It leads to confusion (open a repo > open a .md expecting Markdown > it's actually a PHP program) and it makes life insanely harder for tools that work using file extensions (ex: editor tooling, eslint, prettier etc).

I think other tools that support this made a mistake and we shouldn't fall into this. CMSes should adapt.

**Reply by wassfila** [reply]
in my point of view, a .md file should stick to .md content, it is a misunderstanding to think of this otherwise.
my original intension was not to fake the extension but to be honest with the .md content, which lacks the capability to "augment" components.
I personally vote for a unique markdown standard and extension, and all tooling specific extensions (e.g. components augmentation) are to be abstracted from the user who should not care. mdx pipeline happen to allow md components to be augmented. What for in an Astro project to keep .astro concept and its compiler aside and to start rehyping components with javascript rehype plugins, given that markdown is one of the flare features of Astro and of static content, it deserves some features to allow its community to grow. I'll post separately a more thorough motivation with examples.

At the end of the day the cool thing about this feature is that it is capable to sattisfy both divergent opinions about it
* those who like it, or are desperate as out of options and need it => they can enable it
* those who are fundamentally against it => they can ignore it and not use it

**Comment by wassfila** [comment]
I'm personally not a fan of mdx and prefer to stick to pure .md standard to allow a growing eco system. My goals are simple
* how to keep a standard cross tool compatible markdown database (vs code, github, astro)
* how to customize markdown with astro components to give a unique rendering experience without altering the original .md files

I got stuck in this, not only me, some of the top support squad too, I opened an SO question that summarises my attemps

https://stackoverflow.com/questions/74919230/how-to-customize-markdown-with-astro-components

But the motivation is here
* for ages the web had plenty of ways how to generate html
* Astro has a revolutionary language and compiler to generate html
* Markdown is a main component of the JAM stack revolution
* Astro is selling SSG as it's leading and differenciating feature
* Markdown is a first class citizen in Astro (out of the box experience reaches a huge community)
* is it possible to customize a markdown component with an .astro component ? no
* when people ask about how to enhance md headings, we need to referr to the classic rehype toolchain (total disconnection from Astro experience)

## current workarounds
* I would like in the astro era to avoid this (basically svg assets and html in js in the main config of astro docs)
    https://github.com/withastro/docs/blob/52bf88ec74e3d01d212808d678320f190be64f76/astro.config.ts#L16

* As a pragmatic person, I wrote a python script that copies all .md files from their source (authors do not even know what mdx is) and I rename .md to .mdx (while I can do this and no one see it, I would not be proud to suggest this to others and whish for a more elegant alternative)

## community posts
the search is really hard, only from recent period

https://discord.com/channels/830184174198718474/1077511972163162182

https://discord.com/channels/830184174198718474/1078754947480424549

https://discord.com/channels/830184174198718474/1075478924420730900

https://discord.com/channels/830184174198718474/1074529716352069693

https://discord.com/channels/830184174198718474/1072255962615926885

https://discord.com/channels/830184174198718474/1056870932305694770

## markdown landscape survey and predictions

this is my survey about "markdown" and beyond (light weight markup) landscape
https://github.com/orgs/MicroWebStacks/discussions/8

not only I wish but I will also recommend, help and push towards markdown standardisation and unification of practices. Multiplicity of files extensions is one of them .md, .mdx, .mmd .mdok. I would distinguish flavors and extensions as being website and engine internals and not splitting the content base into heterogeneous incompatible categories nor being used as a lockin opportunity. Therefore I do not see myself deviating from the the .md extension and from a correct minimilastic githug rendering. The rest, styling and augmentations, is the core business of md renderering engines, why would one prevent itself from profiting of its language technology for such enhancements.

**Reply by bholmesdev** [reply]
Hm, this definitely feels out-of-scope for the proposal above. I'd ask for a separate discussion on this!

**Reply by wassfila** [reply]
I agree, maybe my focus is on "customize md components with .astro" and mdx in .md is just a "workaround" to achieve it. Otherwise I agree with Princesseuh, I would not want to see actual mdx content under .md which would also be allowed by this, but hey we're grownups, I can't judge here.

**Reply by wassfila** [reply]
that separate discussion actually already exist 
How to customize markdown with Astro components ? : https://github.com/withastro/roadmap/discussions/423

I updated it now with my motivation from above and split the non goal not to be overlapping with this one.

**Reply by stereobooster** [reply]
related https://github.com/withastro/roadmap/discussions/769

**Comment by ancaemcken** [comment]
Specifically for allowing `.md` extensions for Markdoc - see my [comment](https://github.com/withastro/roadmap/pull/508#issuecomment-1508145519) and the related PR https://github.com/withastro/astro/pull/6847.

**Comment by chardskarth** [comment]
It's been a year now since this was posted and this is a limiting blocker to my use case.

## My Use Case
My team uses a gitlab wiki and I want to integrate a static astro site to make wiki more user friendly.

But gitlab [doesn't display MDX](https://docs.gitlab.com/ee/user/project/wiki/#create-or-edit-wiki-pages-locally).

 So as a workaround, I want to process ".md" files, which gitlab displays in the wiki (but yea i understand it wont render as beautifully), as "MDX".
